Modules/_ctypes/callproc.c

```diff
@@ -1209,7 +1209,12 @@ PyObject *_ctypes_callproc(PPROC pProc,
         }
     }
 
-    rtype = _ctypes_get_ffi_type(restype);
+    if (restype == Py_None) {
+        rtype = &ffi_type_void;
+    } else {
+        rtype = _ctypes_get_ffi_type(restype);
+    }
```

Comment:
I'm not familiar with ctypes. Would it make sense to check for None in ``_ctypes_get_ffi_type()`` rather than here?

```
ffi_type *_ctypes_get_ffi_type(PyObject *obj)
{
    StgDictObject *dict;
    if (obj == NULL)
        return &ffi_type_sint;
    if (obj == Py_None)
        return &ffi_type_void;
    dict = PyType_stgdict(obj);
    if (dict == NULL)
        return &ffi_type_sint;
...
```

Reply:
I think this is better? I ran into a similar issue the `libffi` port itself: `_ctypes_get_ffi_type` is used for the arguments. Return types need a little bit of special handling, like for instance it doesn't make sense to have an argument of type `ffi_type_void`. I would argue that it would be safest to do something like:
```C
if (obj == Py_None)
	// Set an error, arguments can't have type None!
```
but I'm not that familiar with the ctypes interface -- are we supposed to use `None` if we are not sure about the argument's type? Similarly,
```C
    if (obj == NULL)
        return &ffi_type_sint;
```
looks kind of scary to me. Why did we pass in `NULL` to this function? If this function set exceptions in these weird cases and they were handled as appropriate at the call site, the bug I'm fixing here probably wouldn't have happened in the first place.

Reply:
The `NULL` value is used to indicate *not-set* and is quite pervasive in the ctypes codebase.  Function prototypes are not required to supply argtypes or restype.  It seems `_ctypes_get_ffi_type()` is designed for arguments so the additional checking for `Py_None` outside of it would be correct.